File: Queue/649. DOTA2Senate.cpp
class Solution {
public:
    string helper(int rcharge, int dcharge, int rcount, int dcount, queue<char>& q, string senate){
        cout << rcharge << dcharge << rcount << dcount << senate << endl; 
        for(char c: senate){
            if(rcharge == 0 && dcharge == 0){
                q.push(c);
                if(c == 'R') {
                    rcharge += 1;
                    rcount += 1;
                }
                if(c == 'D') {
                    dcharge += 1;
                    dcount += 1;
                } 
            }else if(rcharge > 0){
                if(c == 'R'){
                    rcharge += 1;
                    rcount += 1;
                    q.push(c);
                }
                if(c == 'D') {
                    rcharge -= 1;
                }
            }else if(dcharge > 0){
                if(c == 'D'){
                    dcharge += 1;
                    dcount += 1; 
                    q.push(c);
                }
                if(c == 'R') {
                    dcharge -= 1;
                } 
            }
        }

        if(rcount == 0 && dcount > 0){
            return "Dire";
        }else if(dcount == 0 && rcount > 0){
            return "Radiant";
        }else{
            queue<char> temp; 
            string s = "";
            while(!q.empty()){
                s += q.front();
                temp.push(q.front());
                q.pop();
            }
            return helper(rcharge, dcharge, 0, 0, q, s);
        }
    }

    string predictPartyVictory(string senate) {
        queue<char> q; 
        return helper(0, 0, 0, 0, q, senate); 
    }
};

File: Queue/649. DOTA2Senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        rq = []
        dq = []
        n = len(senate)
        for i in range(n):
            if senate[i] == "R":
                rq.append(i)
            else:
                dq.append(i)

        while len(rq) > 0 and len(dq) > 0:
            if rq[0] < dq[0]:
                rq.append(rq[0] + n)
                rq.pop(0)
                dq.pop(0)
            else:
                dq.append(dq[0] + n)
                rq.pop(0)
                dq.pop(0)
        
        if len(dq) > 0:
            return "Dire"
        else:
            return "Radiant"
